File: chapter_4_trees_graphs/js/4.2-minimalTree.js
const TreeNode = require("./tree")

/* Creates a binary search tree with minimal height */
const createMinimalBST = arr => {
  return minimalBST(arr, 0, arr.length - 1)
}

/* helper */
const minimalBSTHelper = (arr, start, end) => {
  if (start > end) return null
  const mid = Math.floor((start + end) / 2)
  const node = new TreeNode(arr[mid])
  node.left = minimalBSTHelper(arr, start, mid - 1)
  node.right = minimalBSTHelper(arr, mid + 1, end)
  return node
}

const arr = [1, 4, 6, 8, 9, 12, 19, 21]
const root = createMinimalBST(arr)
preOrderTraversal(root)
/*
8 
4
1
6
12
9
19
21
*/

File: chapter_4_trees_graphs/python/4.2-minimal_tree.py
TreeNode = __import__('tree').TreeNode
pre_order_traversal = __import__('tree').pre_order_traversal


def create_minimal_bst(arr):
    """
    Creates a binary search tree with minimal height
    """
    return create_minimal_bst_helper(arr, 0, len(arr) - 1)


def create_minimal_bst_helper(arr, start, end):
    """
    Helper 
    """
    if start > end:
        return None
    mid = (start + end) // 2
    node = TreeNode(arr[mid])
    node.left = create_minimal_bst_helper(arr, start, mid - 1)
    node.right = create_minimal_bst_helper(arr, mid + 1, end)
    return node


arr = [1, 4, 6, 8, 9, 12, 19, 21]
root = create_minimal_bst(arr)
pre_order_traversal(root)
"""
8
4
1
6
12
9
19
21
"""
